More fixups to blog post (#22)
rust-lang/project-ffi-unwind: blogposts/inside-rust/01-announcement.md

Platform ABIs can define a special kind of unwinding called "forced unwinding."
This type of unwinding operation cannot be stopped prematurely using language
mechanisms such as Rust's `catch_unwind` or C++'s `catch`. (These mechanisms
can sometimes intercept forced unwinds, but they must resume the unwind

>  otherwise the runtime will abort the program).

Reviewer: Looking at the spec, it seems there is no suggestion of automatically aborting:

> A catch-all block may be executed during forced unwinding. For instance, a `longjmp may execute code in a catch(...) during stack unwinding. However, if this happens, unwinding will proceed at the end of the catch-all block, whether or not there is an explicit rethrow. 

....but I suppose a compiler could automatically insert some logic to abort rather than letting the unwind be resumed.

I'm trying to determine whether GCC (and by extension, maybe C++ in general?) have the behavior you're describing, where a non-rethrowing `catch` block would `abort` instead of automatically resuming the unwind. But on my Mac, it seems like I don't see destructors being run whether or not I rethrow the `pthread_exit` exception. In fact, I don't even see a print statement in the `catch` block itself. http://coliru.stacked-crooked.com/a/ee07fca96fe69c47

Author: Basically only glibc on Linux uses forced unwinding for pthread_exit. Even musl on Linux just exits the thread directly without unwinding.

The abort happens because glibc defines the destructor of the exception object passed to _Unwind_ForcedUnwind to abort if the exception object is ever destroyed (which happens if you catch it without rethrowing it).

Reviewer: Ah, that makes a lot of sense. It seems like it's probably worth noting that the statement about `pthread_exit` only applies to glibc on Linux, though.

  * It may, of course, cause programs to abort that used to execute
    successfully. This could occur if a panic would've been caught and
    recovered.
* Changing the ABI (the `"C"` in `extern "C"`) of functions in the libc crate is

>   a breaking change: function pointers of different ABIs have different types.

Reviewer: We consider this to be an unacceptable cost, correct? (This should probably be explicit either way.)

Author: I think so. Unless we want to play around with fancy coercion rules... (but I'd rather not)

Reviewer: Yeah, then let's just say something like:

```
  * We cannot change the ABI (the `"C"` in `extern "C"`) of functions in the libc
    crate, because this would be a breaking change: function pointers of different
    ABIs have different types.
```

generally use `extern "C"` when interacting with C APIs (making sure to avoid
destructors where `longjmp` might be used), and `extern "C unwind"`
when interacting with C++ APIs. The main advantage of this proposal is that
switching between `panic=unwind` and `panic=abort` can never introduce UB, but
may cause your program to abort instead of unwind.

### Proposal 2: Introduce `"C unwind"`, forced unwinding always permitted

This is the same as the previous design, except that when compiled with
`panic=abort`, forced unwinding would *not* be intercepted at `"C unwind"` ABI
boundaries; that is, they would behave normally (though still UB if there are
any destructors), without causing the program to abort. `panic`-unwind and
non-forced foreign exceptions would still cause the program to abort.

### Proposal 3: No new ABI

* `panic=unwind`: unwind behaves normally
* `panic=abort`:
  * `panic`-unwind: does not exist; `panic!` aborts the program
  * foreign unwinding, no destructors: unwind behaves normally
  * foreign unwinding with destructors: undefined behavior

The main advantage of this proposal is its simplicity: there is only one ABI and

> the behavior of `panic=unwind` is identical to that of `-fno-exceptions` in C++.

Reviewer: ```suggestion
the behavior of `panic=abort` is identical to that of `-fno-exceptions` in C++.
```

However this comes with the downside that switching to `panic=abort` may in some
cases introduce UB (though only in unsafe code) if FFI calls unwind through Rust
code.

## Comparison table for the proposed designs

In this table, "UB" stands for "undefined behavior". Some instances of UB can
be detected at runtime, but the code to do so would impose an undesirable
code-size penalty; the group recommends generating such code only for debug
builds. These cases are marked "UB (debug: abort)"

> Note that unwinding through a frame without running its destructors (e.g. because they have been optimized out by `panic=abort`) is always undefined behavior, except if the frame has no destructors.

Reviewer: ```suggestion
Note that unwinding through a frame that has destructors without running those destructors (e.g. because they have been optimized out by `panic=abort`) is always undefined behavior.
```
